src/test/scala/com.snowplowanalytics.maxmind.iplookups/IpLookupsTest.scala

```diff
@@ -173,80 +160,102 @@ object IpLookupsTest {
+class IpLookupsTest extends Specification with Tables with CatsEffect {
 
   "Looking up some IP address locations should match their expected locations" should {
-
-    val mcf: Boolean => String = mc => if (mc) "using" else "without using"
-    val lcf: Int => String =
-      lc => if (lc > 0) "LRU cache sized %s".format(lc) else "no LRU cache"
-    val formatter: (String, Boolean, Int) => String =
-      (ip, mcache, lcache) =>
-        "The IP address %s looked up (%s memory cache and with %s)".format(
-          ip,
-          mcf(mcache),
-          lcf(lcache)
-        )
-
     import IpLookupsTest._
 
     for {
       memCache <- Seq(true, false)
       lruCache <- Seq(0, 1000, 10000)
     } {
-
-      val ioIpLookups = ioIpLookupsFromFiles(memCache, lruCache)
-      val idIpLookups = idIpLookupsFromFiles(memCache, lruCache)
-
       testData foreach { case (ip, expected) =>
-        formatter(ip, memCache, lruCache) should {
-          val ioActual = ioIpLookups.performLookups(ip).unsafeRunSync()
-          val idActual = idIpLookups.performLookups(ip)
-          matchIpLookupResult(ioActual, expected)
-          matchIpLookupResult(idActual, expected)
+        prepareTestFormatter(ip, memCache, lruCache) should {
+          "work for IO monad" in {
+            assertWithFiles[IO](memCache, lruCache, ip, expected)
+          }
+          "work for Id monad" in {
+            assertWithFiles[Id](memCache, lruCache, ip, expected)
+          }
         }
       }
     }
 
-    "providing an invalid ip should fail" in {
-      val ioIpLookups = ioIpLookupsFromFiles(true, 0)
-      val idIpLookups = idIpLookupsFromFiles(true, 0)
-      val ioExpected = IpLookupResult(
-        new UnknownHostException("not: Name or service not known").asLeft.some,
-        new UnknownHostException("not: Name or service not known").asLeft.some,
-        new UnknownHostException("not: Name or service not known").asLeft.some,
-        new UnknownHostException("not: Name or service not known").asLeft.some,
-        new UnknownHostException("not: Name or service not known").asLeft.some,
-        new UnknownHostException("not: Name or service not known").asLeft.some
-      )
-      val idExpected = IpLookupResult(
-        new UnknownHostException("not").asLeft.some,
-        new UnknownHostException("not").asLeft.some,
-        new UnknownHostException("not").asLeft.some,
-        new UnknownHostException("not").asLeft.some,
-        new UnknownHostException("not").asLeft.some,
-        new UnknownHostException("not").asLeft.some
-      )
-      val ioActual = ioIpLookups.performLookups("not").unsafeRunSync()
-      val idActual = idIpLookups.performLookups("not")
-      matchIpLookupResult(ioActual, ioExpected)
-      matchIpLookupResult(idActual, idExpected)
+    "providing an invalid ip" should {
+      "fail for IO monad" in {
+        val expected = IpLookupResult(
+          new UnknownHostException("not: Name or service not known").asLeft.some,
```

Comment:
I know you didn't introduce this repetition but would it be okay to factor it out to some variable instead of repeating it ?